API/Extensions/IdentityServiceExtensions.cs

```diff
@@ -151,6 +124,87 @@ public static IServiceCollection AddIdentityServices(this IServiceCollection ser
+            options.RequireHttpsMetadata = options.Authority.StartsWith("https://");
+
+            options.SignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+            options.ResponseType = OpenIdConnectResponseType.Code;
+            options.CallbackPath = "/signin-oidc";
+            options.SignedOutCallbackPath = "/signout-callback-oidc";
+
+            options.SaveTokens = true;
+            options.GetClaimsFromUserInfoEndpoint = true;
+            options.Scope.Clear();
+            options.Scope.Add("openid");
+            options.Scope.Add("profile");
+            options.Scope.Add("offline_access");
+            options.Scope.Add("roles");
+            options.Scope.Add("email");
+            foreach (var customScope in settings.CustomScopes)
+            {
+                options.Scope.Add(customScope);
+            }
+
+            options.Events = new OpenIdConnectEvents
+            {
+                OnTokenValidated = OidcClaimsPrincipalConverter,
+                OnAuthenticationFailed = ctx =>
+                {
+                    ctx.Response.Redirect("/login?skipAutoLogin=true&error="+Uri.EscapeDataString(ctx.Exception.Message));
+                    ctx.HandleResponse();
+                    return Task.CompletedTask;
+                },
+            };
+        });
+    }
+
+    private static AuthenticationBuilder AddOpenIdConnectCookie(this AuthenticationBuilder builder, IWebHostEnvironment environment)
+    {
+        return builder.AddCookie(CookieAuthenticationDefaults.AuthenticationScheme, options =>
+        {
+            options.ExpireTimeSpan = TimeSpan.FromDays(7);
+            options.SlidingExpiration = true;
+
+            options.Cookie.HttpOnly = true;
+            options.Cookie.IsEssential = true;
+            options.Cookie.MaxAge = TimeSpan.FromDays(7);
+
+            if (environment.IsEnvironment(Environments.Development))
+            {
+                options.Cookie.Domain = null;
+            }
+
+            options.Events = new CookieAuthenticationEvents
+            {
+                OnValidatePrincipal = async ctx =>
+                {
+                    var oidcService = ctx.HttpContext.RequestServices.GetRequiredService<IOidcService>();
+                    await oidcService.RefreshCookieToken(ctx);
+                },
+                OnRedirectToAccessDenied = ctx =>
+                {
+                    ctx.Response.StatusCode = StatusCodes.Status401Unauthorized;
+                    return Task.CompletedTask;
+                },
+                OnRedirectToLogin = ctx =>
+                {
+                    ctx.Response.StatusCode = StatusCodes.Status401Unauthorized;
+                    return Task.CompletedTask;
+                },
+            };
+        });
+    }
+
+    /// <summary>
+    /// Called after the OIDC token has been validated, only called on login. Used to find the user we'll be authenticating against
```

Comment:
Update comment to explain a bit more about usage, because it finds the user but seems like a middleware